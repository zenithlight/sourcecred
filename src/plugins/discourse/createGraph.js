// @flow

import {Graph, type Node, type Edge} from "../../core/graph";
import {type NodeWeight} from "../../core/weights";
import {weightsForDeclaration} from "../../analysis/pluginDeclaration";
import {type WeightedGraph} from "../../core/weightedGraph";
import {
  type PostId,
  type TopicId,
  type CategoryId,
  type Tag,
  type Post,
} from "./fetch";
import {type ReadRepository} from "./mirrorRepository";
import {declaration} from "./declaration";
import {
  type DiscourseReference,
  parseLinks,
  linksToReferences,
} from "./references";
import * as NE from "./nodesAndEdges";
import {type DiscourseConfig} from "./config";
import {likeWeight} from "./weights";

export type GraphUser = {|
  +node: Node,
|};

export type GraphTopic = {|
  +node: Node,
  +hasAuthor: Edge,
|};

export type GraphPost = {|
  +topicContains: Edge,
  +node: Node,
  +hasAuthor: Edge,
  +references: $ReadOnlyArray<Edge>,
  +postReplies: Edge | null,
|};

export type GraphLike = {|
  +createsLike: Edge,
  +likes: Edge,
  +node: Node,
  +weight: NodeWeight,
|};

/**
 * TopicHasLikedPost edges connect a Topic to the posts
 * in that topic that were liked, in proportion to the
 * total like weight of the post in question.
 *
 * See: https://github.com/sourcecred/sourcecred/issues/1896
 */
export type TopicHasLikedPost = {|
  +edge: Edge,
  +weight: number,
|};

export type GraphData = {
  +users: $ReadOnlyArray<Node>,
  +topics: $ReadOnlyArray<GraphTopic>,
  +posts: $ReadOnlyArray<GraphPost>,
  +likes: $ReadOnlyArray<GraphLike>,
  +topicHasLikedPosts: $ReadOnlyArray<TopicHasLikedPost>,
};

// TODO: Make this configurable.
// For details on trust levels:
// https://blog.discourse.org/2018/06/understanding-discourse-trust-levels/
export const DEFAULT_TRUST_LEVEL_TO_WEIGHT = Object.freeze({
  "0": 0,
  // Trust level 1 indicates little engagement (doesn't even require any posts)
  // so I gave them a very small weight.
  "1": 0.1,
  // Trust level 2 in my mind indicates being a "full member" so it feels
  // like a good anchor for a standard weight of 1.
  "2": 1,
  // Trust level 3 requires that you are highly active and have earned lots of
  // likes, so feels trusted enough for some bonus minting.
  "3": 1.25,
  // Trust level 4 means you've been designated as high trust by the admins, so
  // we give a bigger bonus. Could make this even larger (2?)
  "4": 1.5,
});

export function _createGraphData(
  config: DiscourseConfig,
  repo: ReadRepository
): GraphData {
  const {serverUrl} = config;
  const users = repo.users().map((u) => NE.userNode(serverUrl, u.username));

  const topicIdToCategory = new Map<TopicId, CategoryId>();
  const topicIdToTags = new Map<TopicId, $ReadOnlyArray<Tag>>();
  const topics: $ReadOnlyArray<GraphTopic> = repo.topics().map((topic) => {
    const node = NE.topicNode(serverUrl, topic);
    topicIdToCategory.set(topic.id, topic.categoryId);
    topicIdToTags.set(topic.id, topic.tags);
    const hasAuthor = NE.authorsTopicEdge(serverUrl, topic);
    return {node, hasAuthor};
  });

  const findPostInTopic = repo.findPostInTopic.bind(repo);
  const postIdToDescription = new Map();
  const posts: $ReadOnlyArray<GraphPost> = repo.posts().map((post) => {
    const topic = repo.topicById(post.topicId);
    const topicTitle = topic != null ? topic.title : "[unknown topic]";
    const url = serverUrl + "/t/" + post.topicId + "/" + post.indexWithinTopic;
    const description = `[#${post.indexWithinTopic} on ${topicTitle}](${url})`;
    postIdToDescription.set(post.id, description);
    const node = NE.postNode(serverUrl, post, description);
    const hasAuthor = NE.authorsPostEdge(serverUrl, post);

    const references = _createReferenceEdges(serverUrl, post, findPostInTopic);

    let postReplies = null;
    let replyToPostIndex = post.replyToPostIndex;
    if (replyToPostIndex == null && post.indexWithinTopic > 1) {
      // The replyToPostIndex gets set to null if it is actually a reply to
      // the topic's own post.
      replyToPostIndex = 1;
    }
    if (replyToPostIndex != null) {
      const parentId = repo.findPostInTopic(post.topicId, replyToPostIndex);
      if (parentId != null) {
        postReplies = NE.postRepliesEdge(serverUrl, post, parentId);
      }
    }

    const topicContains = NE.topicContainsPostEdge(serverUrl, post);

    return {node, hasAuthor, references, postReplies, topicContains};
  });

  const postIdToLikeWeight = new Map();
  const likes: $ReadOnlyArray<GraphLike> = repo.likes().map((like) => {
    const postDescription =
      postIdToDescription.get(like.postId) || "[unknown post]";
    const node = NE.likeNode(serverUrl, like, postDescription);
    const createsLike = NE.createsLikeEdge(serverUrl, like);
    const likes = NE.likesEdge(serverUrl, like);
    const user = repo.findUser(like.username);
    const post = repo.postById(like.postId);
    const topicId = post ? post.topicId : null;
    let weight = 0;
    if (topicId) {
      const categoryId = topicIdToCategory.get(topicId);
      const tags = topicIdToTags.get(topicId);
      weight = likeWeight(config.weights, user, categoryId, tags);
    } else {
      weight = likeWeight(config.weights, user);
    }
    // Update how much total like weight this post has, so that we can
    // set up a hasLikedPost edge flowing cred from the topic
    const existingWeight = postIdToLikeWeight.get(like.postId) || 0;
    postIdToLikeWeight.set(like.postId, existingWeight + weight);

    return {node, createsLike, likes, weight};
  });

  const topicHasLikedPosts = [];
  for (const [postId, weight] of postIdToLikeWeight.entries()) {
    if (weight === 0) {
      // This could happen if all of the likes were from untrusted users
      continue;
    }
    const post = repo.postById(postId);
    if (post == null) {
      // The like didn't correspond to a valid post--maybe a cache/deletion
      // thing--let's ignore it.
      continue;
    }
    const edge = NE.topicHasLikedPostEdge(serverUrl, post);
    topicHasLikedPosts.push({edge, weight});
  }
  return {users, topics, posts, likes, topicHasLikedPosts};
}

export function _graphFromData({
  users,
  topics,
  posts,
  likes,
  topicHasLikedPosts,
}: GraphData): WeightedGraph {
  const g = new Graph();
  const weights = weightsForDeclaration(declaration);
  for (const user of users) {
    g.addNode(user);
  }
  for (const topic of topics) {
    g.addNode(topic.node);
    g.addEdge(topic.hasAuthor);
  }
  for (const post of posts) {
    g.addNode(post.node);
    g.addEdge(post.topicContains);
    g.addEdge(post.hasAuthor);
    if (post.postReplies != null) {
      g.addEdge(post.postReplies);
    }
    for (const reference of post.references) {
      g.addEdge(reference);
    }
  }
  for (const like of likes) {
    g.addNode(like.node);
    g.addEdge(like.createsLike);
    g.addEdge(like.likes);
    weights.nodeWeights.set(like.node.address, like.weight);
  }

  for (const {edge, weight} of topicHasLikedPosts) {
    g.addEdge(edge);
    weights.edgeWeights.set(edge.address, {forwards: weight, backwards: 0});
  }
  return {graph: g, weights};
}

export function createGraph(
  config: DiscourseConfig,
  repo: ReadRepository
): WeightedGraph {
  const data = _createGraphData(config, repo);
  return _graphFromData(data);
}

export function _createReferenceEdges(
  serverUrl: string,
  post: Post,
  findPostInTopic: (topicId: TopicId, indexWithinTopic: number) => ?PostId,
  // This is available as a helper for testing, so we don't need to construct posts
  // with fake html containing links.
  _manualLinks: ?$ReadOnlyArray<string>
): $ReadOnlyArray<Edge> {
  const links = _manualLinks
    ? _manualLinks
    : parseLinks(post.cooked, serverUrl);
  const references = linksToReferences(links);
  const result = [];
  for (const reference of references) {
    const edge = _referenceEdge(serverUrl, post, reference, findPostInTopic);
    if (edge != null) {
      result.push(edge);
    }
  }
  return result;
}

function _referenceEdge(
  serverUrl: string,
  post: Post,
  reference: DiscourseReference,
  findPostInTopic: (topicId: TopicId, indexWithinTopic: number) => ?PostId
): Edge | null {
  if (
    reference.serverUrl != null &&
    reference.serverUrl.toLowerCase() !== serverUrl.toLowerCase()
  ) {
    // Don't attempt to make cross-instance links for now, since we only
    // load one Discourse forum in a given instance.
    return null;
  }
  switch (reference.type) {
    case "TOPIC": {
      return NE.referencesTopicEdge(serverUrl, post, reference);
    }
    case "POST": {
      const referredPostId = findPostInTopic(
        reference.topicId,
        reference.postIndex
      );
      if (referredPostId == null) {
        // Maybe a bad link, or the post or topic was deleted.
        return null;
      }
      return NE.referencesPostEdge(serverUrl, post, referredPostId);
    }
    case "USER": {
      return NE.referencesUserEdge(serverUrl, post, reference);
    }
    default: {
      throw new Error(`Unexpected reference type: ${(reference.type: empty)}`);
    }
  }
}
