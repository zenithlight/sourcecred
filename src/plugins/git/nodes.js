// @flow

import deepFreeze from "deep-freeze";
import {NodeAddress, type NodeAddressT} from "../../core/graph";
import type {Hash} from "./types";

export opaque type RawAddress: NodeAddressT = NodeAddressT;

const GIT_PREFIX = NodeAddress.fromParts(["sourcecred", "git"]);
export function _gitAddress(...parts: string[]): RawAddress {
  return NodeAddress.append(GIT_PREFIX, ...parts);
}

export const COMMIT_TYPE: "COMMIT" = "COMMIT";

export const Prefix: {|base: NodeAddressT, commit: RawAddress|} = deepFreeze({
  base: GIT_PREFIX,
  commit: _gitAddress(COMMIT_TYPE),
});

export type CommitAddress = {|
  +type: typeof COMMIT_TYPE,
  +hash: Hash,
|};

export type StructuredAddress = CommitAddress;

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${NodeAddress.toString(x)}`);
  }
  if (!NodeAddress.hasPrefix(x, GIT_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_git, _type, ...rest] = NodeAddress.toParts(x);
  const type: $ElementType<StructuredAddress, "type"> = (_type: any);
  switch (type) {
    case "COMMIT": {
      if (rest.length !== 1) throw fail();
      const [hash] = rest;
      return {type: COMMIT_TYPE, hash};
    }
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case COMMIT_TYPE:
      return NodeAddress.append(Prefix.commit, x.hash);
    default:
      throw new Error(`Unexpected type ${(x.type: empty)}`);
  }
}
