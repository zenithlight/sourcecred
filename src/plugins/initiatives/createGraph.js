// @flow

import {
  type Edge,
  type Node,
  type EdgeAddressT,
  type NodeAddressT,
  EdgeAddress,
  NodeAddress,
} from "../../core/graph";
import type {ReferenceDetector, URL} from "../../core/references";
import {type WeightedGraph as WeightedGraphT} from "../../core/weightedGraph";
import * as WeightedGraph from "../../core/weightedGraph";
import type {NodeWeight} from "../../core/weights";
import type {EdgeSpec} from "./edgeSpec";
import {
  type Initiative,
  type InitiativeId,
  type InitiativeRepository,
  addressFromId,
} from "./initiative";
import {
  type NodeEntry,
  type NodeEntryField,
  addressForNodeEntry,
} from "./nodeEntry";
import {
  dependsOnEdgeType,
  referencesEdgeType,
  contributesToEdgeType,
  championsEdgeType,
  contributesToEntryEdgeType,
} from "./declaration";
import {initiativeFileURL} from "./initiativeFile";

function initiativeAddress(initiative: Initiative): NodeAddressT {
  return addressFromId(initiative.id);
}

function initiativeNode(initiative: Initiative): Node {
  const address = initiativeAddress(initiative);
  const url = initiativeFileURL(address);
  return {
    address,
    timestampMs: initiative.timestampMs,
    description:
      url == null ? initiative.title : `[${initiative.title}](${url})`,
  };
}

function nodeFromEntry(
  entry: NodeEntry,
  parentId: InitiativeId,
  field: NodeEntryField
): Node {
  const address = addressForNodeEntry(field, parentId, entry.key);
  return {
    address,
    timestampMs: entry.timestampMs,
    description: entry.title,
  };
}

export function initiativeWeight(initiative: Initiative): ?NodeWeight {
  if (!initiative.weight) return;
  return initiative.completed
    ? initiative.weight.complete
    : initiative.weight.incomplete;
}

type EdgeFactoryT = (initiative: Initiative, other: NodeAddressT) => Edge;

function edgeFactory(
  prefix: EdgeAddressT,
  fromInitiative: boolean
): EdgeFactoryT {
  return (initiative: Initiative, other: NodeAddressT): Edge => {
    const iAddr = initiativeAddress(initiative);
    const src = fromInitiative ? iAddr : other;
    const dst = fromInitiative ? other : iAddr;
    return {
      address: EdgeAddress.append(
        prefix,
        ...NodeAddress.toParts(initiativeAddress(initiative)),
        ...NodeAddress.toParts(other)
      ),
      timestampMs: initiative.timestampMs,
      src,
      dst,
    };
  };
}

const depedencyEdge = edgeFactory(dependsOnEdgeType.prefix, true);
const referenceEdge = edgeFactory(referencesEdgeType.prefix, true);
const contributionEdge = edgeFactory(contributesToEdgeType.prefix, false);
const championEdge = edgeFactory(championsEdgeType.prefix, false);

export function createWeightedGraph(
  repo: InitiativeRepository,
  refs: ReferenceDetector
): WeightedGraphT {
  const wg = WeightedGraph.empty();
  const {graph, weights} = wg;

  for (const initiative of repo.initiatives()) {
    // Adds the Initiative node.
    const node = initiativeNode(initiative);
    const weight = initiativeWeight(initiative);
    graph.addNode(node);
    if (weight) {
      weights.nodeWeights.set(node.address, weight);
    }

    // Generic approach to adding edges when the reference detector has a hit.
    const urlHandler = (
      urls: $ReadOnlyArray<URL>,
      createEdge: EdgeFactoryT
    ) => {
      for (const url of urls) {
        const addr = refs.addressFromUrl(url);
        if (!addr) {
          console.warn(`initiative ${initiative.title}: no address for ${url}`);
          continue;
        }
        graph.addEdge(createEdge(initiative, addr));
      }
    };

    // Generic approach to handling EdgeSpecs.
    const edgeSpecHandler = (
      {urls, entries}: EdgeSpec,
      createEdge: EdgeFactoryT,
      field: NodeEntryField
    ) => {
      // Delegate handling the URLs.
      urlHandler(urls, createEdge);

      for (const entry of entries) {
        // Add the NodeEntry contribution itself to the graph.
        const entryNode = nodeFromEntry(entry, initiative.id, field);
        graph.addNode(entryNode);
        graph.addEdge(createEdge(initiative, entryNode.address));
        if (entry.weight != null) {
          weights.nodeWeights.set(entryNode.address, entry.weight);
        }

        // Add edges to the contributors.
        for (const contributor of entry.contributors) {
          const addr = refs.addressFromUrl(contributor);
          if (!addr) {
            console.warn(
              `entry ${entry.title}: no address for contributor ${contributor}`
            );
            continue;
          }
          graph.addEdge({
            address: EdgeAddress.append(
              contributesToEntryEdgeType.prefix,
              ...NodeAddress.toParts(entryNode.address),
              ...NodeAddress.toParts(addr)
            ),
            timestampMs: entry.timestampMs,
            src: addr,
            dst: entryNode.address,
          });
        }
      }
    };

    // Maps the edge types to it's fields.
    edgeSpecHandler(initiative.dependencies, depedencyEdge, "DEPENDENCY");
    edgeSpecHandler(initiative.references, referenceEdge, "REFERENCE");
    edgeSpecHandler(initiative.contributions, contributionEdge, "CONTRIBUTION");
    urlHandler(initiative.champions, championEdge);
  }

  return wg;
}
