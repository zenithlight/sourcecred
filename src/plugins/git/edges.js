// @flow

import deepFreeze from "deep-freeze";
import {
  type Edge,
  type EdgeAddressT,
  EdgeAddress,
  NodeAddress,
} from "../../core/graph";
import * as GitNode from "./nodes";
import {type TimestampMs} from "../../util/timestamp";

export opaque type RawAddress: EdgeAddressT = EdgeAddressT;

export const HAS_TREE_TYPE: "HAS_TREE" = "HAS_TREE";
export const HAS_PARENT_TYPE: "HAS_PARENT" = "HAS_PARENT";
export const INCLUDES_TYPE: "INCLUDES" = "INCLUDES";
export const BECOMES_TYPE: "BECOMES" = "BECOMES";
export const HAS_CONTENTS_TYPE: "HAS_CONTENTS" = "HAS_CONTENTS";

const GIT_PREFIX = EdgeAddress.fromParts(["sourcecred", "git"]);
function gitEdgeAddress(...parts: string[]): RawAddress {
  return EdgeAddress.append(GIT_PREFIX, ...parts);
}

export const Prefix: {|base: EdgeAddressT, hasParent: RawAddress|} = deepFreeze(
  {
    base: GIT_PREFIX,
    hasParent: gitEdgeAddress(HAS_PARENT_TYPE),
  }
);

export type HasParentAddress = {|
  type: typeof HAS_PARENT_TYPE,
  child: GitNode.CommitAddress,
  parent: GitNode.CommitAddress,
|};

export type StructuredAddress = HasParentAddress;

export const createEdge: {|
  hasParent: (
    child: GitNode.CommitAddress,
    parent: GitNode.CommitAddress,
    timestampMs: TimestampMs
  ) => Edge,
|} = deepFreeze({
  hasParent: (
    child: GitNode.CommitAddress,
    parent: GitNode.CommitAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: HAS_PARENT_TYPE, child, parent}),
    src: GitNode.toRaw(child),
    dst: GitNode.toRaw(parent),
    timestampMs,
  }),
});

const NODE_PREFIX_LENGTH = NodeAddress.toParts(GitNode._gitAddress()).length;

function lengthEncode(x: GitNode.RawAddress): $ReadOnlyArray<string> {
  const baseParts = NodeAddress.toParts(x).slice(NODE_PREFIX_LENGTH);
  return [String(baseParts.length), ...baseParts];
}
function lengthDecode(
  x: $ReadOnlyArray<string>,
  fail: () => Error
): {|+parts: $ReadOnlyArray<string>, +rest: $ReadOnlyArray<string>|} {
  if (x.length === 0) {
    // Not length-encoded.
    throw fail();
  }
  const [lengthString, ...allParts] = x;
  const length = parseInt(lengthString, 10);
  if (isNaN(length)) {
    throw fail();
  }
  if (length > allParts.length) {
    // Not enough elements.
    throw fail();
  }
  return {parts: allParts.slice(0, length), rest: allParts.slice(length)};
}
function multiLengthDecode(x: $ReadOnlyArray<string>, fail: () => Error) {
  let remaining = x;
  const partses = [];
  while (remaining.length > 0) {
    const {parts, rest} = lengthDecode(remaining, fail);
    partses.push(parts);
    remaining = rest;
  }
  return partses;
}

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${EdgeAddress.toString(x)}`);
  }
  if (!EdgeAddress.hasPrefix(x, GIT_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_git, _type, ...rest] = EdgeAddress.toParts(x);
  const type: $ElementType<StructuredAddress, "type"> = (_type: any);
  switch (type) {
    case "HAS_PARENT": {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 2) throw fail();
      const [childParts, parentParts] = parts;
      const child: GitNode.CommitAddress = (GitNode.fromRaw(
        GitNode._gitAddress(...childParts)
      ): any);
      const parent: GitNode.CommitAddress = (GitNode.fromRaw(
        GitNode._gitAddress(...parentParts)
      ): any);
      return {type: HAS_PARENT_TYPE, child, parent};
    }
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case HAS_PARENT_TYPE:
      return EdgeAddress.append(
        Prefix.hasParent,
        ...lengthEncode(GitNode.toRaw(x.child)),
        ...lengthEncode(GitNode.toRaw(x.parent))
      );
    default:
      throw new Error((x.type: empty));
  }
}
