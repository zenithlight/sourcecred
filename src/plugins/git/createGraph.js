// @flow

import {Graph} from "../../core/graph";

import * as GT from "./types";
import * as GN from "./nodes";
import * as GE from "./edges";

export function createGraph(repository: GT.Repository): Graph {
  const creator = new GraphCreator();
  creator.addRepository(repository);
  return creator.graph;
}

class GraphCreator {
  +graph: Graph;

  constructor() {
    this.graph = new Graph();
  }

  addRepository(repository: GT.Repository) {
    for (const commitHash of Object.keys(repository.commits)) {
      this.addCommit(repository.commits[commitHash]);
    }
  }

  addCommit(commit: GT.Commit) {
    const structuredAddress: GN.CommitAddress = {
      type: GN.COMMIT_TYPE,
      hash: commit.hash,
    };
    const address = GN.toRaw(structuredAddress);
    const description = commit.hash;
    this.graph.addNode({address, description, timestampMs: commit.createdAt});
    for (const parentHash of commit.parentHashes) {
      const parent: GN.CommitAddress = {type: GN.COMMIT_TYPE, hash: parentHash};
      this.graph.addEdge(
        GE.createEdge.hasParent(structuredAddress, parent, commit.createdAt)
      );
    }
  }
}
