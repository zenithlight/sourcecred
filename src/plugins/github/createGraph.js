// @flow

import * as NullUtil from "../../util/null";
import {Graph} from "../../core/graph";
import {type WeightedGraph} from "../../core/weightedGraph";
import {type WeightsT, empty as emptyWeights} from "../../core/weights";
import * as GitNode from "../git/nodes";
import * as N from "./nodes";
import * as R from "./relationalView";
import {createEdge} from "./edges";
import {ReactionContent$Values as Reactions} from "./graphqlTypes";

export function createGraph(view: R.RelationalView): WeightedGraph {
  const creator = new GraphCreator();
  creator.addData(view);
  return {graph: creator.graph, weights: creator.weights};
}

class GraphCreator {
  graph: Graph;
  weights: WeightsT;

  constructor() {
    this.graph = new Graph();
    this.weights = emptyWeights();
  }

  addData(view: R.RelationalView) {
    for (const entity of view.entities()) {
      const address = N.toRaw(entity.address());
      this.graph.addNode({
        address,
        description: entity.description(),
        timestampMs: entity.timestampMs(),
      });
    }

    for (const child of view.childEntities()) {
      this.addHasParent(child);
    }

    for (const authored of view.authoredEntities()) {
      this.addAuthors(authored);
    }

    for (const pull of view.pulls()) {
      const commit = pull.mergedAs();
      if (commit != null) {
        const commitTimestamp = NullUtil.get(view.commit(commit)).timestampMs();
        this.graph.addEdge(
          createEdge.mergedAs(pull.address(), commit, commitTimestamp)
        );
      } else {
        const addr = N.toRaw(pull.address());
        // Un-merged PRs do not mint cred.
        this.weights.nodeWeights.set(addr, 0);
      }
    }

    for (const commit of view.commits()) {
      const gitCommitAddress: GitNode.CommitAddress = {
        type: GitNode.COMMIT_TYPE,
        hash: commit.hash(),
      };
      this.graph.addEdge(
        createEdge.correspondsToCommit(
          commit.address(),
          gitCommitAddress,
          commit.timestampMs()
        )
      );
    }

    for (const referrer of view.textContentEntities()) {
      for (const referent of referrer.references()) {
        this.graph.addEdge(
          createEdge.references(
            referrer.address(),
            referent.address(),
            referrer.timestampMs()
          )
        );
      }
    }

    for (const reactable of view.reactableEntities()) {
      for (const {content, user, timestampMs} of reactable.reactions()) {
        // We only support unambiguously positive reactions for now
        if (
          content === Reactions.THUMBS_UP ||
          content === Reactions.HEART ||
          content === Reactions.HOORAY ||
          content === Reactions.ROCKET
        ) {
          this.graph.addEdge(
            createEdge.reacts(content, user, reactable.address(), timestampMs)
          );
        }
      }
    }
  }

  addAuthors(entity: R.AuthoredEntity) {
    for (const author of entity.authors()) {
      this.graph.addEdge(
        createEdge.authors(
          author.address(),
          entity.address(),
          entity.timestampMs()
        )
      );
    }
  }

  addHasParent(child: R.Issue | R.Pull | R.Comment | R.Review) {
    this.graph.addEdge(
      createEdge.hasParent(
        child.address(),
        child.parent().address(),
        child.timestampMs()
      )
    );
  }
}
