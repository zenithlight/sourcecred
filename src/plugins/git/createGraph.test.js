// @flow

import cloneDeep from "lodash.clonedeep";

import {createGraph} from "./createGraph";
import {Prefix as NodePrefix} from "./nodes";
import {Prefix as EdgePrefix} from "./edges";
import {NodeAddress, EdgeAddress} from "../../core/graph";

const makeData = () => cloneDeep(require("./example/example-git"));

// Disabled while the Git plugin is inactive
describe.skip("plugins/git/createGraph", () => {
  describe("createGraph", () => {
    it("processes a simple repository", () => {
      expect(createGraph(makeData())).toMatchSnapshot();
    });

    it("only has commit nodes and has_parent edges", () => {
      const graph = createGraph(makeData());
      for (const {address} of graph.nodes()) {
        if (!NodeAddress.hasPrefix(address, NodePrefix.commit)) {
          throw new Error("Found non-commit node");
        }
      }
      for (const {address} of graph.edges({showDangling: true})) {
        if (!EdgeAddress.hasPrefix(address, EdgePrefix.hasParent)) {
          throw new Error("Found non-has-parent edge");
        }
      }
    });
  });
});
