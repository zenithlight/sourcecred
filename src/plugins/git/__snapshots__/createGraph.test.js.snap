// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/git/createGraph createGraph processes a simple repository 1`] = `
Array [
  Object {
    "type": "sourcecred/graph",
    "version": "0.8.0",
  },
  Object {
    "edges": Array [
      Object {
        "address": Array [
          "sourcecred",
          "git",
          "HAS_PARENT",
          "2",
          "COMMIT",
          "3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f",
          "2",
          "COMMIT",
          "69c5aad50eec8f2a0a07c988c3b283a6490eb45b",
        ],
        "dstIndex": 1,
        "srcIndex": 0,
        "timestampMs": 981173106000,
      },
      Object {
        "address": Array [
          "sourcecred",
          "git",
          "HAS_PARENT",
          "2",
          "COMMIT",
          "69c5aad50eec8f2a0a07c988c3b283a6490eb45b",
          "2",
          "COMMIT",
          "e8b7a8f19701cd5a25e4a097d513ead60e5f8bcc",
        ],
        "dstIndex": 7,
        "srcIndex": 1,
        "timestampMs": 981173106000,
      },
      Object {
        "address": Array [
          "sourcecred",
          "git",
          "HAS_PARENT",
          "2",
          "COMMIT",
          "8d287c3bfbf8455ef30187bf5153ffc1b6eef268",
          "2",
          "COMMIT",
          "c08ee3a4edea384d5291ffcbf06724a13ed72325",
        ],
        "dstIndex": 3,
        "srcIndex": 2,
        "timestampMs": 981173106000,
      },
      Object {
        "address": Array [
          "sourcecred",
          "git",
          "HAS_PARENT",
          "2",
          "COMMIT",
          "c08ee3a4edea384d5291ffcbf06724a13ed72325",
          "2",
          "COMMIT",
          "c2b51945e7457546912a8ce158ed9d294558d294",
        ],
        "dstIndex": 4,
        "srcIndex": 3,
        "timestampMs": 981173106000,
      },
      Object {
        "address": Array [
          "sourcecred",
          "git",
          "HAS_PARENT",
          "2",
          "COMMIT",
          "c90f6424017f787bbbaf22e4082a01355546f7e3",
          "2",
          "COMMIT",
          "3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f",
        ],
        "dstIndex": 0,
        "srcIndex": 5,
        "timestampMs": 981173106000,
      },
      Object {
        "address": Array [
          "sourcecred",
          "git",
          "HAS_PARENT",
          "2",
          "COMMIT",
          "d160cca97611e9dfed642522ad44408d0292e8ea",
          "2",
          "COMMIT",
          "8d287c3bfbf8455ef30187bf5153ffc1b6eef268",
        ],
        "dstIndex": 2,
        "srcIndex": 6,
        "timestampMs": 981173106000,
      },
      Object {
        "address": Array [
          "sourcecred",
          "git",
          "HAS_PARENT",
          "2",
          "COMMIT",
          "e8b7a8f19701cd5a25e4a097d513ead60e5f8bcc",
          "2",
          "COMMIT",
          "d160cca97611e9dfed642522ad44408d0292e8ea",
        ],
        "dstIndex": 6,
        "srcIndex": 7,
        "timestampMs": 981173106000,
      },
    ],
    "nodes": Array [
      Object {
        "description": "3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f",
        "index": 0,
        "timestampMs": 981173106000,
      },
      Object {
        "description": "69c5aad50eec8f2a0a07c988c3b283a6490eb45b",
        "index": 1,
        "timestampMs": 981173106000,
      },
      Object {
        "description": "8d287c3bfbf8455ef30187bf5153ffc1b6eef268",
        "index": 2,
        "timestampMs": 981173106000,
      },
      Object {
        "description": "c08ee3a4edea384d5291ffcbf06724a13ed72325",
        "index": 3,
        "timestampMs": 981173106000,
      },
      Object {
        "description": "c2b51945e7457546912a8ce158ed9d294558d294",
        "index": 4,
        "timestampMs": 981173106000,
      },
      Object {
        "description": "c90f6424017f787bbbaf22e4082a01355546f7e3",
        "index": 5,
        "timestampMs": 981173106000,
      },
      Object {
        "description": "d160cca97611e9dfed642522ad44408d0292e8ea",
        "index": 6,
        "timestampMs": 981173106000,
      },
      Object {
        "description": "e8b7a8f19701cd5a25e4a097d513ead60e5f8bcc",
        "index": 7,
        "timestampMs": 981173106000,
      },
    ],
    "sortedNodeAddresses": Array [
      Array [
        "sourcecred",
        "git",
        "COMMIT",
        "3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f",
      ],
      Array [
        "sourcecred",
        "git",
        "COMMIT",
        "69c5aad50eec8f2a0a07c988c3b283a6490eb45b",
      ],
      Array [
        "sourcecred",
        "git",
        "COMMIT",
        "8d287c3bfbf8455ef30187bf5153ffc1b6eef268",
      ],
      Array [
        "sourcecred",
        "git",
        "COMMIT",
        "c08ee3a4edea384d5291ffcbf06724a13ed72325",
      ],
      Array [
        "sourcecred",
        "git",
        "COMMIT",
        "c2b51945e7457546912a8ce158ed9d294558d294",
      ],
      Array [
        "sourcecred",
        "git",
        "COMMIT",
        "c90f6424017f787bbbaf22e4082a01355546f7e3",
      ],
      Array [
        "sourcecred",
        "git",
        "COMMIT",
        "d160cca97611e9dfed642522ad44408d0292e8ea",
      ],
      Array [
        "sourcecred",
        "git",
        "COMMIT",
        "e8b7a8f19701cd5a25e4a097d513ead60e5f8bcc",
      ],
    ],
  },
]
`;
